tools/cmaketobzl/cmaketobzl.go

```diff
@@ -0,0 +1,420 @@
+		e.v.SetParent(key, strings.Join(args[0:len(args)-1], ";"))
+		// When setting CACHE variables, the option can occur in either position -3 or -4 due to the optional FORCE parameter.
+	case len(args) >= 3 && args[len(args)-3] == "CACHE", len(args) >= 4 && args[len(args)-4] == "CACHE":
+		log.Println("Ignoring set of CACHE variable: ", key)
+	default:
+		e.v.Set(key, strings.Join(args, ";"))
+	}
+}
+
+// unsetVariable unsets the value of the variable designated by the remained, following the rules of
+// https://cmake.org/cmake/help/v3.3/command/set.html#command:unset
+func (e *eval) unsetVariable(args []string) {
+	switch {
+	case len(args) == 0:
+		log.Println("Cannot unset a variable without a name")
+	case len(args) == 1:
+		e.v.Set(args[0], "")
+	case len(args) == 2 && args[1] == "PARENT_SCOPE":
+		e.v.SetParent(args[0], "")
+	case len(args) == 2 && args[1] == "CACHE":
+		log.Println("Ignoring unset of CACHE variable: ", args[0])
+	default:
+		log.Println("Ignoring invalid unset command")
+	}
+}
+
+// AddSubdirectory recurses into the directory specified by dirpath and evaluates the CMakeLists.txt contained therein.
+func (e *eval) AddSubdirectory(dirpath string) error {
+	if err := e.enterDirectory(dirpath); err != nil {
+		return err
+	}
+	file, err := e.parseFile(path.Join(path.Join(e.path...), "CMakeLists.txt"))
+	if err != nil {
+		return err
+	}
+
+	cmds := commandList(file.Commands)
+	dispatch := e.dispatch
+	for {
+		dispatch, err := dispatch(&cmds)
+		if err != nil {
+			return err
+		}
+		if len(cmds) == 0 || dispatch == nil {
+			break
+		}
+
+	}
+	return e.exitDirectory(dirpath)
+}
+
+// CurrentDirectory returns the project-rooted path currently being traversed.
+func (e *eval) CurrentDirectory() string {
+	return "/" + path.Join(e.path[1:]...)
+}
+
+// enterDirectory pushes a new directory onto the stack, setting up necessary state, etc.
+func (e *eval) enterDirectory(dirpath string) error {
+	if err := e.w.PushDirectory(dirpath); err != nil {
+		return err
+	}
+	e.v.Push()
+	e.path = append(e.path, dirpath)
+	e.v.Set("CMAKE_CURRENT_SOURCE_DIR", e.CurrentDirectory())
+	e.v.Set("CMAKE_CURRENT_BINARY_DIR", e.CurrentDirectory())
+	return nil
+}
+
+// exitDirectory pops the most recently entered directory off the stack.
+func (e *eval) exitDirectory(path string) error {
+	e.v.Pop()
+	e.path = e.path[:len(e.path)-1]
+	tail, err := e.w.PopDirectory()
+	if tail != path {
+		return fmt.Errorf("unexpected directory state %v != %v", tail, path)
+	}
+	return err
+}
+
+// PrintCommand writes the given command to the configured StarlarkWriter.
+func (e *eval) PrintCommand(command *ast.CommandInvocation) error {
+	return e.w.WriteCommand(string(command.Name), command.Arguments.Eval(e.v)...)
+}
+
+// Path is a slice of string segments, representing a filesystem path.
+type Path []string
+
+// Split cleans and splits the /-delimited filesystem path.
+func NewPath(s string) Path {
+	p := strings.Split(filepath.ToSlash(path.Clean(s)), "/")
```

Comment:
I believe you want to invert the composition here: `path.Clean(filepath.ToSlash(s))`; otherwise you're cleaning a path in native format in slash form. Alternatively: `filepath.ToSlash(filepath.Clean(s))` works.

Reply:
Switched to `filepath.ToSlash(filepath.Clean(s))` and moved to a separate package.